torch/masked/maskedtensor/functions.py

```diff
@@ -0,0 +1,150 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+
+import math
+
+import torch
+from torch.nn.functional import dropout, linear
+
+from .matmul import masked_bmm
+
+__all__ = ['multi_head_attention_forward']
+
+def _in_projection_packed(q, k, v, w, b):
+    w_q, w_k, w_v = w.chunk(3)
+    if b is not None:
+        raise ValueError("b must be None")
+    b_q = b_k = b_v = None
+    return linear(q, w_q, b_q), linear(k, w_k, b_k), linear(v, w_v, b_v)
+
+
+def _scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=0.0):
+    B, Nt, E = q.shape
+    q = q / math.sqrt(E)
+    # (B, Nt, E) x (B, E, Ns) -> (B, Nt, Ns)
+    attn = masked_bmm(q, k.transpose(-2, -1), attn_mask)
+    attn = torch.nn.functional.softmax(attn, dim=-1)
+    if dropout_p > 0.0:
+        attn = dropout(attn, p=dropout_p)
+    # (B, Nt, Ns) x (B, Ns, E) -> (B, Nt, E)
+    output = torch.bmm(attn, v)
+    return output, attn
+
+
+def multi_head_attention_forward(
+    query,
+    key,
+    value,
+    embed_dim_to_check,
+    num_heads,
+    in_proj_weight,
+    in_proj_bias,
+    bias_k,
+    bias_v,
+    add_zero_attn,
+    dropout_p,
+    out_proj_weight,
+    out_proj_bias,
+    training,
+    key_padding_mask,
+    need_weights,
+    attn_mask,
+    use_separate_proj_weight,
+    q_proj_weight,
+    k_proj_weight,
+    v_proj_weight,
+    static_k,
+    static_v,
+    average_attn_weights,
+):
+    # set up shape vars
+    tgt_len, bsz, embed_dim = query.shape
+    src_len, _, _ = key.shape
+
+    if embed_dim != embed_dim_to_check:
+        raise ValueError(f"Was expecting embedding dimension of {embed_dim_to_check}, but got {embed_dim}")
+
+    if isinstance(embed_dim, torch.Tensor):
```

Comment:
just curious - what prevents using the implementation of `torch.nn.functional` ?